README.md
# procps_ptree - shows inforamtion of executed process and  task

<a href="https://scan.coverity.com/projects/rjaan-procps_ptree">
  <img alt="Coverity Scan Build Status"
       src="https://scan.coverity.com/projects/20648/badge.svg"/>
</a>

  _Copyright (C) 2012-2020 Andrey Y. Rzhavskov_

  **procps_ptree** is modified source code of project [procps-ng](https://gitlab.com/procps-ng/procps) and  demonstrates some 
  features which library procps provides in its own APIs.  This program is written in learning purposes only.

  Home's page on Rusian language:
	[procps_ptree on narod.ru](http://rjaan.narod.ru/projects/procps_ptree.html)

  About procps-ng
        [procps-ng](https://gitlab.com/procps-ng/procps/-/blob/master/README.md)

Comments are welcome.

 _Andrey Y. Rzhavskov <rjaan@yandex.ru>_



CHANGELOG.md
**This file is no longer supported.  Please see git log for the most
up to date list about contributions.**

> $ git clone https://github.com/rjaan/procps_ptree.git
> $ cd procps_ptree
> $ git log
> 

*-- for old changelog --* 

#For version 1.1

**2012-07-23**  _Andrey Rzhavskov  rjaan@yandex.ru_

* file CHANGELOG was added;
* added the pointer-to-pointer 'tab' free in type proc_data_t

#For version 1.2

**2020-03-17**  _Andrey Rzhavskov  rjaan@yandex.ru_

* some requirments are provided to registrate on GitHub .

#For version 1.3

**2020-03-18**  _Andrey Rzhavskov  rjaan@yandex.ru_

* re-oreder project to it meet all of criteria from [coverity.com](https://scan.coverity.com/faq) that they were adopted from the [Open Source Initiative](http://opensource.org/docs/definition.php)





src/output.c
 /* 
   the file output.c contains functions that output information about executed 
		     processes and threads. They are using standard I/O streams 
		     such as stdout, stderr. 

   Copyright (C) 2012  Andrey Y. Rzhavskov  <mailto:rjaan@yandex.ru>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA 

*/

#include "procps_ptree.h"

#include <proc/readproc.h>
#include <proc/escape.h>
#include <proc/sysinfo.h>


PRINT_INFO(pid);
PRINT_INFO(pid);
PRINT_INFO(ppid);		
PRINT_INFO(stat);
PRINT_INFO(user);
PRINT_INFO(pcpu);
PRINT_INFO(pmem);
PRINT_INFO(stime);			
PRINT_INFO(time);
PRINT_INFO(comm);
PRINT_INFO(thread);

typedef void ( * pr_handler_t ) ( char *restrict const outbuf, 
			       		      const proc_t *restrict const pp );

typedef struct pr_list {
   pr_handler_t	 pr;
   const char 	*fmt;
   const char 	*header;	
   char		 sep_ch[2];
}pr_list_t; 


static const pr_list_t pr_list_proc[]= {
{ pr_pid,  "%6s",  "PID",{ SEPORATOR_SYMBOL_SPACE, SEPORATOR_SYMBOL_SEMI } },
{ pr_ppid, "%6s",  "PPID",{ SEPORATOR_SYMBOL_SPACE, SEPORATOR_SYMBOL_SEMI } },		
{ pr_stat, "%7s",  "STAT",{ SEPORATOR_SYMBOL_SPACE, SEPORATOR_SYMBOL_SEMI } },
{ pr_user, "%11s", "USER",{ SEPORATOR_SYMBOL_SPACE, SEPORATOR_SYMBOL_SEMI } },
{ pr_pcpu, "%4s",  "%%CPU",{ SEPORATOR_SYMBOL_SPACE, SEPORATOR_SYMBOL_SEMI } },
{ pr_pmem, "%4s",  "%%MEM",{ SEPORATOR_SYMBOL_SPACE, SEPORATOR_SYMBOL_SEMI } },
{ pr_stime,"%16s", "START_TIME",{ SEPORATOR_SYMBOL_SPACE, SEPORATOR_SYMBOL_SEMI } },			
{ pr_time, "%12s", "TIME",{ SEPORATOR_SYMBOL_SPACE, SEPORATOR_SYMBOL_SEMI } },
{ NULL, NULL, "",{ 0, 0 } }	  	
};

void show_header ( void )
{
    const pr_list_t *fmt_list = pr_list_proc;

    if ( glvar_to_show_only_pids ) {
         return;
    }
 
    fputc ( '\n', stdout );
    fflush(stdout);
    while ( fmt_list && fmt_list->pr ) {
          fprintf ( stdout, fmt_list->fmt, fmt_list->header );
          fputc ( fmt_list->sep_ch[(glvar_to_use_csv_format ? 1 : 0)], stdout );		
	  fmt_list++;
    }
    fprintf ( stdout, " COMMAND");
    fflush(stdout);	
    fputc ( '\n', stdout );
    fflush(stdout);
}

void show_one_proc( const proc_t *p, int level, int is_show )
{
    int i = level;
    const pr_list_t *fmt_list = pr_list_proc;  
    static char outbuf[PATH_MAX] = { 0 };

    if(!is_show) return;
	
    while ( fmt_list && fmt_list->pr ) {
    	bzero( outbuf, PATH_MAX );
	fmt_list->pr( outbuf, p );  
	fprintf ( stdout, fmt_list->fmt, outbuf );
	if ( glvar_to_show_only_pids ) {
	  fputc ( '\n', stdout ); 
	  fflush(stdout);
	  return; 
        }
	fputc ( fmt_list->sep_ch[(glvar_to_use_csv_format ? 1 : 0)], stdout );		
	fmt_list++;
    }

    if ( !glvar_to_use_csv_format )    
       while ( i-- ) {
	  int space = 3;
	  while ( space-- )
	      fputc ( ' ', stdout );
       }

    pr_comm ( outbuf, p );  	
    fprintf ( stdout, "%s\n", outbuf );
    fflush(stdout);	

}

#define ROWLENGTH  240 /* row length */

/******************************************************************************
 * Output formated functins						      *	
 ******************************************************************************/
 static void pr_pid ( char *restrict const outbuf, const proc_t *restrict const pp ) {
   if ( pp->tid != pp->tgid ) {
      pr_thread ( outbuf, pp );	
      return;
   }
   (void)snprintf(outbuf, ROWLENGTH, "%6u", pp->tgid);
 }

 static void pr_ppid(char *restrict const outbuf, const proc_t *restrict const pp) {
  if ( pp->tid != pp->tgid ) { 
     (void)snprintf(outbuf, ROWLENGTH, "%6u", pp->tgid);
     return;	  
  } 	
  (void)snprintf(outbuf, ROWLENGTH, "%6u", pp->ppid);
 }

 static void pr_comm(char *restrict const outbuf, const proc_t *restrict const pp) {
   int rightward = 16;	
   (void)escape_command ( outbuf, pp, ROWLENGTH, &rightward, ESC_DEFUNCT );
 }


 static void pr_user(char *restrict const outbuf, const proc_t *restrict const pp){
    int rightward = 10;

    if ( !glvar_dont_resolve_name )  
       (void)escape_str(outbuf, pp->euser, ROWLENGTH, &rightward);
    else
       (void)snprintf ( outbuf, ROWLENGTH, "%6u", pp->euid );
 }

 static void pr_pcpu(char *restrict const outbuf, const proc_t *restrict const pp){
  /* seconds of process life */
   unsigned long long seconds    = glvar_seconds_since_boot - pp->start_time / Hertz; 
  /* jiffies used by this process */
   unsigned long long total_jiffies = pp->utime + pp->stime;
  /* scaled %cpu, 999 means 99.9% */
   unsigned int pcpu;

   if ( !seconds ) { /* to except the division by zero error */
     (void)snprintf(outbuf, ROWLENGTH, "%2u.0", 0 );  	
      return;	
   }	

   pcpu = (total_jiffies * 1000ULL / Hertz) / seconds;

   if (pcpu > 999U ) {
    (void)snprintf(outbuf, ROWLENGTH, "99.9" );
    return; 
   }
   (void)snprintf(outbuf, ROWLENGTH, "%2u.%u", pcpu/10U, pcpu%10U);
 }

 static void pr_pmem(char *restrict const outbuf, const proc_t *restrict const pp){
/* scaled %mem, 999 means 99.9% */
  unsigned long pmem;
  if( kb_main_total ) { /* to except the division by zero error */
   (void)snprintf(outbuf, ROWLENGTH, "DZ.ERR" ); 	
  }	
  pmem = pp->vm_rss * 1000ULL / kb_main_total;	
  if ( pmem > 999) pmem = 999;

  (void)snprintf(outbuf, ROWLENGTH, "%2u.%u", (unsigned)(pmem/10), (unsigned)(pmem%10)); 
}

static void pr_stime(char *restrict const outbuf, const proc_t *restrict const pp){

 struct tm *tm; /* date and time to broken-down time*/
 int our_yday;  /* our year day */
 int our_year;   /* our year */

 const char *fmt = NULL; /* Объявить формат вывода для strftime() */

/* seconds since 1970 of boot system */
 time_t  seconds_since_1970_of_boot = glvar_seconds_since_1970 - 
						       glvar_seconds_since_boot;
/* seconds since 1970 of start process */
 time_t  seconds_since_1970_of_life = seconds_since_1970_of_boot + 
						       (pp->start_time / Hertz);

 tm = localtime(&glvar_seconds_since_1970);
 our_yday = tm->tm_yday;
 our_year = tm->tm_year;

 tm = localtime ( &seconds_since_1970_of_life ); 
 if( our_yday != tm->tm_yday ) fmt = "%m-%d %H:%M:%S";  /* mm-dd HH:MM:SS */
 if( our_year != tm->tm_year ) fmt = "%Y-%m-%d %H:%M:%S";  /* YY-mm-dd HH:MM:SS */
 if( fmt == NULL ) fmt = "%H:%M:%S"; /* HH:MM:SS*/

 (void)strftime(outbuf, 42, fmt, tm  ); 
}

static void pr_time ( char *restrict const outbuf, const proc_t *restrict const pp ) {

    unsigned int  dd_chars, dd, mm, hh, ss;	
    unsigned long cpu_seconds;

    if ( !Hertz ) {
      snprintf(outbuf, ROWLENGTH, "DIV:ZERO:ERR" );	
      return; 
    }
	
    cpu_seconds = (pp->utime + pp->stime) / Hertz;

    ss = cpu_seconds % 60;

    cpu_seconds /= 60;	   
    mm = cpu_seconds % 60;

    cpu_seconds /= 60;
    hh = cpu_seconds % 24; 

    dd = cpu_seconds / 24;
    dd_chars  = ( dd ? snprintf(outbuf, 240, "%u-", dd) : 0 );

    (void)snprintf ( outbuf+dd_chars, ROWLENGTH, "%02u:%02u:%02u", hh, mm, ss );
}

static void pr_stat(char *restrict const outbuf, const proc_t *restrict const pp ) {

    int offset = 0;

    *(outbuf+offset++) = pp->state; /* set one of state-codes:
					 D    Uninterruptible sleep (usually IO)
 					 R    Running or runnable (on run queue)
 					 S    Interruptible sleep (waiting for an event to complete)
 					 T    Stopped, either by a job control signal or because it is being traced.
 					 X    dead (should never be seen)
 					 Z    Defunct ("zombie") process, terminated but not reaped by its parent */
/* add state-code to state-codes set */
    if(pp->nice < 0)                  *(outbuf+offset++) = '<'; /* high-priority (not nice to other users) */
    if(pp->nice > 0)                  *(outbuf+offset++) = 'N'; /* low-priority (nice to other users) */
    if(pp->vm_lock)                   *(outbuf+offset++) = 'L'; /* pages locked into memory (for real-time and custom IO) */
    if(pp->session == pp->tgid)       *(outbuf+offset++) = 's'; /*  session leader */
    if(pp->nlwp > 1)                  *(outbuf+offset++) = 'l'; /*  multi-threaded */
    if(pp->pgrp == pp->tpgid)         *(outbuf+offset++) = '+'; /*  in foreground process group */

    *(outbuf+offset) = '\0'; /* end of line */
}

static void pr_thread ( char *restrict const outbuf, const proc_t *restrict const pp ) {
  (void)snprintf(outbuf, ROWLENGTH, "%6u", pp->tid);
}

/*eof*/



src/parser.c
/* 
   File parser.c contains some functions to parse options into command line.  
 
   Copyright (C) 2012  Andrey Y. Rzhavskov  <mailto:rjaan@yandex.ru>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA 

 */
 #include "procps_ptree.h"

 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include <stdarg.h>
 #include <time.h>

 #include <sys/types.h>
 #include <sys/param.h>

 #include <unistd.h>

 typedef struct jump_table_struct {

	const char *name; /* a name of long option of command-line */	
        const void *jump; /* a pointer on constant of jump-label */
	
 }jump_table_struct_t;

static int compare_jump_table_structs(const void *ch_p, const void *array_p){

   const jump_table_struct_t*  a = (const jump_table_struct_t*)ch_p;
   const jump_table_struct_t*  b = (const jump_table_struct_t*)array_p;

  return strcmp ( a->name, b->name );
}

static int globalvar_by_get_option_name ( const char buf[],
					        	int *glvar_help,
			  				int *glvar_version,
						        int *glvar_loose_task,
			  				int *glvar_csv_format,
			  				int *glvar_only_pids,
			  				int *glvar_dont_resolve_name,
			  				int *glvar_with_pid )
{
 jump_table_struct_t      findme = { buf, NULL };
 jump_table_struct_t	 *found; /* a jump found labels */
 static const jump_table_struct_t jump_table[] = { /* the table of jump labels */
	{ "key04",     &&case_longHelp },    /* long option --help */
	{ "key07",     &&case_longVersion },  /* long option --version */
	{ "key10",     &&case_longLooseTask },  /* long option --loose-task */
	{ "key13",     &&case_longShowWithPid  }, /* long option --show-with-pid */
	{ "key14",     &&case_longShowOnlyPids }, /* long option --show-only-pids*/
	{ "key15",     &&case_longCsvFormat }, /* long option --csv-format-file  */
	{ "key16",     &&case_longNotResolvedName }, /* long option --not-resolve-name */
        { "key56",     &&case_longVersion }, /* short option -V same as long option --version */
        { "key63",     &&case_longCsvFormat }, /* short option -c same as long option --csv-format-file */
        { "key68",     &&case_longHelp }, /* short option -h same as long option --help */
        { "key6C",     &&case_longLooseTask }, /* short option -l same as long option --loose-task */
        { "key6E",     &&case_longNotResolvedName }, /* short option -n same as long option --not-resolve-name */
        { "key70",     &&case_longShowWithPid }, /* short option -p same as long option --show-with-pid */
        { "key73",     &&case_longShowOnlyPids } /* short option -p same as long option --show-only-pids */
 };
 const int jump_table_count =  /* calculating a number of jump labels into the table   */
			   sizeof(jump_table)/sizeof(jump_table_struct_t);
#ifdef __DEBUG__
 printf("%s: find \'%s\' in jump_table_count=[%d]\n", __FUNCTION__, 
						      findme.name , 
						      jump_table_count );   
#endif /*__DEBUG__*/
 
/* search jump label*/
 found = bsearch ( &findme, jump_table, jump_table_count,
      		      sizeof(jump_table_struct_t), compare_jump_table_structs );

 if(!found) {
    perror("bsearch");	
    return -1; 
 } 	

 goto *(found->jump);

 case_longHelp:
 	if(glvar_help) *glvar_help = 1;
	goto case_Success;

 case_longVersion:
 	if(glvar_help) *glvar_version = 1;
	goto case_Success;

 case_longCsvFormat:
	if(glvar_csv_format) *glvar_csv_format = 1; 
	goto case_Success;
 case_longLooseTask:
	if(glvar_loose_task) *glvar_loose_task = 1;
	goto case_Success;
 case_longShowOnlyPids:
	if(glvar_only_pids) *glvar_only_pids = 1;
	goto case_Success;
 case_longNotResolvedName:
        if(glvar_dont_resolve_name) *glvar_dont_resolve_name = 1;
	goto case_Success;
  case_longShowWithPid:
	if(glvar_with_pid) *glvar_with_pid          = 1;
	goto case_Success;
 case_Success:
	return 0;
}


#define OPTION_NOTFOUND()\
 printf ( "%s: invalid option -- \'%s\'\n", *argv, *p_argv_go )

#define ARGUMENT_NOTFOUND(NAME)\
 printf ( "%s: \'%s\' option required argument\n", *argv, NAME )


/*
   cmdline_opts_parser() parses  command-line options  that were  passed    into 
			 2nd argument and initializes global variables that were
			 passed since 3rd argument.
 	
		     	 On success, this function  returns  a number of options 
		         were recognized. Otherwise, it returns negative value.
*/
int cmdline_opts_parser ( const int argc, char * const *argv, 
			  int *to_show_help,
			  int *to_show_version,
			  int *to_use_loose_task,
			  int *to_use_csv_format,
			  int *to_show_only_pids,
			  int *dont_resolve_name,
			  int *to_show_with_pid ) 
 {
    va_list  ap;
    int      n_parsed_opts  	   = 0;      
    int      to_go          	   = argc-1;   
    char     * const *p_argv_go    = argv+1; 
    char     is_wait_arg    	   = 0;
    int      to_show_with_pid_old  = -1;

    if ( to_show_help )      *to_show_help      = 0;
    if ( to_show_version )   *to_show_version   = 0;	
    if ( to_use_loose_task ) *to_use_loose_task = 0;
    if ( to_use_csv_format ) *to_use_csv_format = 0;
    if ( to_show_only_pids ) *to_show_only_pids = 0;
    if ( dont_resolve_name ) *dont_resolve_name = 0;
    if ( to_show_with_pid )  *to_show_with_pid  = to_show_with_pid_old;

    for ( ; to_go && *p_argv_go; p_argv_go++, to_go-- ) {
        char         optkey[16] = { 0 };         
        char 		*ptr    = *p_argv_go;
	while ( *ptr++ ) 
	      ;
        if ( ( ptr - *p_argv_go ) <= 3 ) {
            const char *ptr = *p_argv_go;
            snprintf( optkey, 16, "key%02X", *(ptr+1) );       
        }
	if ( ( ptr - *p_argv_go ) > 3 )	{	
	   snprintf( optkey, 16, "key%02d", (ptr-*p_argv_go)-3 );
        }
	if( globalvar_by_get_option_name ( optkey, 
						 to_show_help,
					         to_show_version,
					  	 to_use_loose_task,
			  			 to_use_csv_format,
			  			 to_show_only_pids,
			  			 dont_resolve_name,
			  			 to_show_with_pid  ) < 0 ) {
	     OPTION_NOTFOUND();
	     n_parsed_opts = -1;
	     goto out;
	 }
         if ( *to_show_with_pid != to_show_with_pid_old ) {
           int boolean_value = (int)(*(++p_argv_go) == NULL);
           const char *ptr = *p_argv_go;
           for ( ; !boolean_value && ptr && *ptr ; ptr++ ){
             if(isdigit( *ptr )) continue;
             boolean_value++;
           }
           if ( boolean_value ) {
		ARGUMENT_NOTFOUND("-p,--show_with_pid");
                *to_show_with_pid = 0;
	        n_parsed_opts = -1;
	        goto out;
           }
	   *to_show_with_pid = atoi(*p_argv_go);		
	   to_show_with_pid_old = *to_show_with_pid;	
         }
         n_parsed_opts++;
    }	   
out:
    return  EXIT_SUCCESS;
 }

/*eof*/



src/procps_ptree.h
/* 
   The file procps_ptree.h contains some function prototypes and macros .  
 
   Copyright (C) 2012  Andrey Y. Rzhavskov  <mailto:rjaan@yandex.ru>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA 
*/
#ifndef __PROCPS_PTREE_H
#define __PROCPS_PTREE_H	1

#include <config.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include <time.h>

#include <sys/types.h>
#include <sys/param.h>

#include <unistd.h>

extern int cmdline_opts_parser ( const int argc, char * const *argv, 
			  int *to_show_help,
			  int *to_show_version,
			  int *to_use_loose_task,
			  int *to_use_csv_format,
			  int *to_show_only_pids,
			  int *dont_resolve_name,
			  int *to_show_with_pid ); /* parses cmd-line 
								options and 
						    		initialize lobal vars */


#define HAVE_INVALID_OPTION  "invalid option"


#define VERSION_BANER()\
puts ("The " PACKAGE " version " VERSION ".\n"\
      "To report a bug please contact " PACKAGE_BUGREPORT ".");

#define VERSION_SHORTBANER()\
puts ("The " PACKAGE " version " VERSION ".");

extern unsigned long   	glvar_seconds_since_boot;
extern time_t		glvar_seconds_since_1970;
extern int 		glvar_to_use_csv_format;
extern int 		glvar_to_show_only_pids;
extern int		glvar_dont_resolve_name;


#define PRINT_INFO(NAME) \
static void pr_ ## NAME (char *restrict const outbuf,\
					const proc_t *restrict const pp);

#define SEPORATOR_SYMBOL_SPACE 		' '
#define SEPORATOR_SYMBOL_SEMI  		';'
#define SEPORATOR_SYMBOL_RETURN  	'\n'

#define SEPORATOR_DEFAULT		NULL

#include <proc/readproc.h>

extern void show_header ( void );
extern void show_one_proc ( const proc_t *p, int level, int is_show );

#endif /*__PROCPS_PTREE_H*/



src/main.c
/* 
   The file main.c contains a program shows hierarchy of processes along with 
                   their threads and is modified source code of procps-ng and 
	           demonstrates some facilities of library procps.  
 
   Copyright (C) 2012  Andrey Y. Rzhavskov  <mailto:rjaan@yandex.ru>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.

 */

#include "procps_ptree.h"


#include <proc/readproc.h>
#include <proc/escape.h>
#include <proc/sysinfo.h>

static int to_show_help;
static int to_show_version;
static int to_use_loose_task;
static int to_show_with_pid;

int glvar_to_use_csv_format;
int glvar_to_show_only_pids;
int glvar_dont_resolve_name;

#define PRINT_VARIABLE(VARIABLE)\
	printf( "%s(): %s = %d\n" , __FUNCTION__, # VARIABLE, VARIABLE )

static void help_message ( void );
static void  global_vars_init ( void );

static  proc_t	 	 **processes;

static proc_data_t* procps_tab_open ( void );
static void procps_tab_close ( proc_data_t  *pd_p );
static int procps_plist_sort ( proc_data_t  *pd_p );

static void show_tree ( const int self, const int n, const int level, 
					  const int have_sibling, int is_fill ); 

int
main ( int argc, char **argv )
{
  int 		   retval  = EXIT_SUCCESS;
  proc_data_t*     pd_p  = NULL;
  int              parent; 
  
  if ( (retval=cmdline_opts_parser ( argc, argv, 
				  	   &to_show_help,
			          	   &to_show_version,
					   &to_use_loose_task,
					   &glvar_to_use_csv_format,
					   &glvar_to_show_only_pids,
					   &glvar_dont_resolve_name,
					   &to_show_with_pid   )) < 0 ) {
     help_message();
     exit(retval);
  }
 
  if ( to_show_help ) {
       help_message();
       exit(0);
  }
  if ( to_show_version ) {
       VERSION_BANER();
       exit(0);
  }

  if( !glvar_to_show_only_pids )
  		VERSION_SHORTBANER();

  global_vars_init ();
 
  pd_p = procps_tab_open ();
  if ( pd_p == NULL ) {
     fprintf(stderr,"%s: can\'t open procps!\n", *argv );
     return EXIT_FAILURE;
  }

  if ( procps_plist_sort ( pd_p ) ) {
        fprintf(stderr, "can\'t sort a proclist\n");
        retval = EXIT_FAILURE;	 
        goto close_proctab; 	
  }

#ifdef __DEBUG__
  PRINT_VARIABLE(to_show_version);
  PRINT_VARIABLE(to_use_loose_task);
  PRINT_VARIABLE(glvar_to_use_csv_format);
  PRINT_VARIABLE(glvar_to_show_only_pids);
  PRINT_VARIABLE(glvar_dont_resolve_name);
  PRINT_VARIABLE(to_show_with_pid);
#endif /*__DEBUG__*/

  show_header();

  parent = pd_p->n;
  while ( parent-- ) {
     int subparent = pd_p->n;
     while ( subparent-- ) {	
         if ( processes[subparent]->XXXID == processes[parent]->ppid ) {
	    goto not_root;
         }
     }
     show_tree ( parent, pd_p->n, 0, 0, 0 ); 
not_root:
	  ;
  }
     
close_proctab:

  procps_tab_close ( pd_p );

  if ( !glvar_to_show_only_pids )
     if( retval == EXIT_FAILURE ) puts ("Failure");
     else puts ("Success");

  return retval;
}

/*
[--help] [--loose-task][--csv-format][--show-only-pid][--not-resolve-name][--show-with-pid root_pid or -p root_pid]
*/

static void help_message ( void )
{
   printf ("%s - \
shows hierarchy of process along with  their tasks\n", PACKAGE_STRING );
  printf ("Usage: %s [OPTION]\n", PACKAGE_STRING );
  printf ("\
Options:\n\
  -h, --help                 display this help and exit\n\
  -V, --version              output version information and exit\n\
  -l, --loose-task           show threads as if they were processes\n\
  -c, --csv-format-file      output processes and threads in the CSV format\n\
  -s, --show-only-pids       only output pids of processes and threads\n\
  -n, --not-resolve-name     don\'t resolve a user name at user id.\n\
  -p, --show-with-pid        shows a tree of processes and threads from this pid\n\
");

}

/* some defines and user types to be required by sorting proctab */
typedef int (*sr_t)(const proc_t* P, const proc_t* Q);

typedef struct sort_node {

  struct sort_node    *next;
  int (*sr)(const proc_t* P, const proc_t* Q); /* sort function */

} sort_node_t;

#define CMP_SMALL(NAME) \
static int sr_ ## NAME (const proc_t* P, const proc_t* Q) { \
    return (int)(P->NAME) - (int)(Q->NAME); \
}

#define CMP_INT(NAME) \
static int sr_ ## NAME (const proc_t* P, const proc_t* Q) { \
    if (P->NAME < Q->NAME) return -1; \
    if (P->NAME > Q->NAME) return  1; \
    return 0; \
}

static  int  		flag;
static  PROCTAB 	*restrict proctab_p;

unsigned long   	glvar_seconds_since_boot;
time_t			glvar_seconds_since_1970;

static  sort_node_t  	  *sort_list;

static void  global_vars_init ( void )
{

  flag = ( PROC_FILLMEM | PROC_FILLSTATUS | PROC_FILLSTAT | 
	   PROC_FILLCOM | PROC_FILLUSR );

  if ( to_use_loose_task ) {
     flag |= PROC_LOOSE_TASKS;
  }
   
  if ( glvar_dont_resolve_name ) {
     flag &= ~PROC_FILLUSR; 
  }

  glvar_seconds_since_boot    = uptime(0,0);
  glvar_seconds_since_1970    = time(NULL);

  proctab_p		      = NULL;
  sort_list                   = NULL;
   
  meminfo ();
} 

static int want_this_proc_nop_cb ( proc_t *dummy ) {
  (void*)dummy;
  return 1;
}

static int want_this_task_nop_cb ( proc_t *dummy ) {
  (void*)dummy;
  return 1;

} 

static proc_data_t* procps_tab_open ( void )
{
  proc_data_t	 *pd_p;

  proctab_p = openproc ( flag );   
  if ( !proctab_p ) {
    fprintf(stderr, "failure: can not access /proc.\n");
    return NULL;
  }

  if( flag & PROC_LOOSE_TASKS ) {
    pd_p = readproctab2 ( want_this_proc_nop_cb, 
			  want_this_task_nop_cb,  proctab_p ); 	
  }else{
    pd_p = readproctab2 ( want_this_proc_nop_cb, 
			  		   NULL,  proctab_p ); 		
  }

  processes = pd_p->tab;

  return pd_p;
}

static void procps_tab_close ( proc_data_t	 *pd_p )
{   
   unsigned long      address_min = 0UL;
   proc_t             *data_p     = NULL;
   unsigned int       ntab        = pd_p->n;
   proc_t             **tab_pp    = pd_p ? pd_p->tab : NULL;

   while ( tab_pp && ntab-- ) {
     proc_t   *p = tab_pp[ntab];   
     if ( !address_min ) {
	   address_min = (unsigned long)p;
     }
     if( (unsigned long)p < address_min ) {
	    address_min = (unsigned long)p;  
	    data_p = p;	
     }

     if( !(flag & PROC_LOOSE_TASKS) && p->cmdline )
		free((void*)*p->cmdline);
     if (p->environ)
		free((void*)*p->environ);
   }
   
   if ( data_p ) {
      pd_p->n = 0;
      pd_p->ntask = 0;
      pd_p->nproc = 0;
      free ( data_p );	
      free ( pd_p );
   }
   
   closeproc( proctab_p ); 
}

CMP_SMALL(ppid);
CMP_INT(start_time);

static int compare_two_procs(const void *a, const void *b ) {

   sort_node_t *tmp_list = sort_list;

   while ( tmp_list != (sort_node_t *)NULL ) {
     int result = (*tmp_list->sr)(*(const proc_t *const*)a, *(const proc_t *const*)b);
     if( result ) return result;	
     tmp_list = tmp_list->next;
   }
   return 0; /* no conclusion */
}


static int procps_plist_sort ( proc_data_t 	*pd_p )
{
  sort_node_t  *tmp_list = sort_list;	

  tmp_list = malloc( sizeof(sort_node_t));  /* ppid */
  if ( !tmp_list ) {	
    return -1;	
  }	
  tmp_list->sr   = sr_ppid;
  tmp_list->next = sort_list; /*  0xdeadbeef */
  sort_list      = tmp_list; 
  
  tmp_list = malloc ( sizeof(sort_node_t) );  /* start_time */
  if ( !tmp_list ) {	
    free( tmp_list );	
    return -1;	
  }	 
  tmp_list->sr   = sr_start_time;
  tmp_list->next = sort_list;
  sort_list      = tmp_list; 
  
  qsort ( pd_p->tab, pd_p->n, sizeof(proc_t*), compare_two_procs );

  if ( sort_list->next ) { 
     free ( sort_list->next ); 
  }
  free ( sort_list );
  
  return 0;
}

/**
  * recurse-show procs tree
  * 
  **/
static void show_tree ( const int self, const int n, const int level, const int have_sibling, int is_fill ) 
{
    int 	i=0;
    proc_t 	task; 
    int 	to_fill = is_fill;
    
    if( !to_fill ) {
      if ( to_show_with_pid == -1 ) to_fill = 1;
      else
      if ( to_show_with_pid == processes[self]->tgid ) to_fill = 1;
    }

    show_one_proc( processes[self], level, to_fill );

    if ( !to_use_loose_task )
      while ( readtask ( proctab_p, processes[self], &task ) ) {
	     if( task.tid == task.tgid ) continue; /* this is my process */

	     show_one_proc( &task, level+1, to_fill );	
      }
   
    i=0;	
    for (; i < n; i++ ) {
	int have_brother = 0;
        if( processes[self]->XXXID != processes[i]->ppid ) continue;
	if( (i+1) < n ) {
   	  if(processes[i+1]->ppid == processes[i]->ppid ) have_brother++;
	}
	show_tree ( i, n, level+1, have_brother, to_fill );  
    }  
	
}


/*eof*/

